xbmc/services/ServiceManager.cpp

```diff
@@ -0,0 +1,49 @@
+#include "ServiceManager.h"
+
+namespace SERVICES
+{
+
+CServiceManager& CServiceManager::GetInstance()
+{
+  static CServiceManager manager;
+  return manager;
+}
+
+bool CServiceManager::RegisterService(std::shared_ptr<IService> service, size_t level)
+{
+  if (!service->StartService())
+    return false;
+
+  CSingleLock lock(*this);
+  if (m_services.size() < level+1)
+    m_services.resize(level+1);
+  m_services[level].emplace_back(std::move(service));
+
+  return true;
+}
+
+bool CServiceManager::TearDown(size_t level)
+{
+  if (level >= m_services.size())
+    return false;
+
+  CSingleLock lock(*this);
+  size_t ok = 0;
+  for (auto it = m_services[level].rbegin(); it != m_services[level].rend(); ++it, ++ok)
+  {
+    if (!(*it)->StopService())
```

Comment:
the lock pretends that stopping the service here is secure but it is not. it stops the service while a client may hold a pointer to an interface.

Since you dropped the logic for bringing up the serviecs from this class, you can drop start/stop completely. An implicit call to StartService in RegisterService seems not right anyway.

Stopping services is not synchronous. We should send signals, requestStop, to the registered services and wait for acc. 
In addition ServiceMangager could track if it handed away interfaces. Then only request stop if the service is not used by any clients.

Reply:
It is not pretending anything. It protects m_services. Ref counting takes care of any holders and it is the service' responsibity to handle (ignore) calls after stop. I thought i was verbose enough in pr desc; the start/stop thing incl making it async will be handled properly later but it takes a lot of additional refactoring and i am not ready to target that yet.

Reply:
Controlling and visualizing dependencies between services was the idea for current service mamanger. By taking this away there is not much left that justifies the name "manager".  Maybe I miss more things you have planned with this refactor. But maybe ServceRegistry is a better fit for what it does.

Reply:
??? It is a 100% direct replacement for the current code except without all the boilerplating.

Reply:
Current ServiceManager has Init method(s) that starts the services in required order. By looking at it, you get an idea of how services depend on each other. I named it manager becuase it was supposed to manage the lifetime of the services and gain control over the sequence how they need to be started.
You said that you moved this part into CApp but it's not clear to me where you intend to move it finally.